### PR TITLE
add support for datasketches js udafs using wasm libraries

### PR DESCRIPTION
**Pull request by nikunjbhartia**
Adding support for 3 datasketches sketch types : 
1) Theta sketch 
2) KLL sketch 
3) Tuple sketch 

Total functions: 
- 9 Aggregate functions ( JS UDAFs) 
- 7 scalar functions  ( JS UDF) 

Solution approach: Datasketch cpp libraries are compiled to WASM libraries using emscripten toolchain and loaded in BQ JS UDAFs. using docker.io emscripten image in build file

Note: currently unit tests are executed using `dataform test` which do no support UDAFs or constant arguments. 
Hence, added test cases for 7 scalar functions only 

There is also a datasketches-cpp directory which could be added as a submodule, committed all the files for now, lets discuss during review.

### CODE REVIEW

**Comment by danieldeleo** [comment]
I've submitted https://github.com/GoogleCloudPlatform/bigquery-utils/pull/425 to enable unit testing for UDAFs. Once it's merged, please pull the changes and add the remaining 9 test cases for the UDAFs

**Comment by nikunjbhartia** [comment]
Thanks for adding support for UDAF unit testing. I have added unit test for 1 UDAF which did not have any const arg in its signature. Rest of the 8 UDAFs have const args and hence can't be added yet. 

Current framework uses positional args which would fail for functions having const args. we need to add additional support for passing constant args  which are defined as "NOT AGGREGATE" in UDAF function signature.

**Comment by danieldeleo** [comment]
@nikunjbhartia you should be good to go for the rest of the 8 UDAFs now that https://github.com/GoogleCloudPlatform/bigquery-utils/pull/426 is merged

**Comment by nikunjbhartia** [comment]
Thanks @danieldeleo for the quick fix. Made following updates: 
- Added unit tests for rest of the 8 UDAFs with const args
- Removed datasketches-cpp folder 
- Added a build task for cloning the above repo

File: udfs/cloudbuild.yaml [hunk 1]
```diff
@@ -53,6 +53,27 @@ steps:
...
+  id: build_wasm_libs
+  dir: datasketches
+  entrypoint: bash
+  args:
+  - '-c'
+  - |
+    for dir in kll-sketch theta-sketch tuple-sketch; do
+      cd $dir && make clean && make all
+      cd ..
+    done
```
Comment (danieldeleo) [inline udfs/cloudbuild.yaml:76]:
```suggestion
#########################################################################
  # Clone datasketches-cpp github repo as dependency for building WASM libraries
  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
  #########################################################################
- name: docker.io/emscripten/emsdk
  id: build_wasm_libs
  dir: datasketches
  entrypoint: bash
  args:
  - '-c'
  - |
    git clone https://github.com/apache/datasketches-cpp.git
    for dir in kll-sketch theta-sketch tuple-sketch; do
      cd $dir && make clean && make all
      cd ..
    done
```

File: udfs/cloudbuild.yaml [hunk 2]
```diff
@@ -53,6 +53,22 @@ steps:
   args:
     - run-script
     - build-all-libs
+#########################################################################
+  # Clone datasketches-cpp github repo as dependency for building WASM libraries
+  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
+  #########################################################################
```
Comment (danieldeleo) [inline udfs/cloudbuild.yaml:59]:
```suggestion
 
 #########################################################################
  # Clone datasketches-cpp github repo as dependency for building WASM libraries
  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
  #########################################################################
```

File: udfs/community/kll_sketch_float64.sqlx [hunk 3]
```diff
@@ -0,0 +1,170 @@
+config { hasOutput: true }
+/*
+ * Copyright 2021 Google LLC
```
Comment (danieldeleo) [inline udfs/community/kll_sketch_float64.sqlx:3]:
```suggestion
 * Copyright 2024 Google LLC
```

Comment (danieldeleo) [reply udfs/community/kll_sketch_float64.sqlx:3]: Update all other new files to be 2024 as well

File: udfs/datasketches/README.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+from UNNEST(GENERATE_ARRAY(1,1000000)) as x1,
+     UNNEST(GENERATE_ARRAY(4,6)) as x2;
+```
+2.Groups on a certain key and aggregates theta_sketches per key
+```sql
+CREATE OR REPLACE TABLE `$BQ_PROJECT.$BQ_DATASET`.theta_agg_by_key_set_a AS
+select
+    group_key,
+    count(*) as total_count,
+    `$BQ_PROJECT.$BQ_DATASET`.theta_sketch_int64(x,14) as theta_sketch
```
Comment (danieldeleo) [inline udfs/datasketches/README.md:84]: Now that we're publishing all the UDFs and UDAFs in the bqutil.fn dataset, let's save the user some manual work and just have all the examples in this README reference the public UDFs

File: udfs/datasketches/kll-sketch/crypto.js [hunk 5]
```diff
@@ -0,0 +1 @@
+var crypto = { getRandomValues: (array) => { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };
```
Comment (danieldeleo) [inline udfs/datasketches/kll-sketch/crypto.js:1]: Please add license header

File: udfs/datasketches/kll-sketch/kll_sketch.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
+#include <stdio.h>
```
Comment (danieldeleo) [inline udfs/datasketches/kll-sketch/kll_sketch.cpp:1]: Please add license header here and to all other source files

Comment (nikunjbhartia) [reply udfs/datasketches/kll-sketch/kll_sketch.cpp:16]: Updated

File: udfs/datasketches/kll-sketch/crypto.js (hunk 5 above)
Comment (nikunjbhartia) [reply udfs/datasketches/kll-sketch/crypto.js:16]: Updated

File: udfs/datasketches/README.md (hunk 4 above)
Comment (nikunjbhartia) [reply udfs/datasketches/README.md]: Updated

File: udfs/community/kll_sketch_float64.sqlx (hunk 3 above)
Comment (nikunjbhartia) [reply udfs/community/kll_sketch_float64.sqlx]: Updated

**Review by danieldeleo — APPROVED** [review]

**Pull request merged**